### PR TITLE
Protect attributes

### LINKED ISSUE #38: Protect writing to attributes that should be accessed through .edit or shouldn't be writable at all
The Dataset and Variable classes have a list of mutable and immutable attributes. You can access them as attributes to be read, like "dataset.name" but, right now it is also possible to do dataset.name = "Something", that should be forbidden since we don't actually save that operation to Crunch. The only way to edit those attributes is by doing a dataset.edit(name="Something else"). Immutable attributes are not editable even with that syntax.

So the idea would be to have a class from which Dataset, Variable, (and potentially Category and others) will inherit that would implement that protection by looking at the _MUTABLE_ATTRIBUTES and _IMMUTABLE_ATTRIBUTES attributes lists.

### PR DESCRIPTION
**Pull request by skriems**
resolves #38 for
- Dataset
- Variable
- Category

### CODE REVIEW

File: scrunch/datasets.py [hunk 1]
```diff
@@ -680,7 +681,7 @@ def __readonly__(self, *args, **kwargs):
     del __readonly__
 
 
-class Dataset(object):
+class Dataset(ReadOnly, object):
```
Comment (jjdelc) [inline scrunch/datasets.py]: We don't need `object` here anymore

File: scrunch/datasets.py [hunk 2]
```diff
@@ -1456,15 +1454,11 @@ def __repr__(self):
...
-    _categories = None
-
     @property
     def categories(self):
         if self.resource.body['type'] not in self.CATEGORICAL_TYPES:
             raise TypeError("Variable of type %s do not have categories" % self.resource.body.type)
-        if self._categories is None:
-            self._categories = CategoryList(self.resource)
-        return self._categories
+        return CategoryList._from(self.resource)
```
Comment (jjdelc) [inline scrunch/datasets.py]: Watch for a regression here, I had a test that made sure that you didn't return a new CategoryList instance each time you accessed the property. That `if` cached the instance once and reused it. With your change we always instantiate and init a new one. Is this deliberate?

File: scrunch/helpers.py [hunk 3]
```diff
@@ -6,6 +6,26 @@
...
     from urllib.parse import urljoin
 
+
+class ReadOnly(object):
+    """
+    class for protecting undesired writes to attributes
+    """
+    def __init__(self, resource):
+        object.__setattr__(self, "resource", resource)
+        object.__setattr__(self, "session", resource.session)
```
Comment (jjdelc) [inline scrunch/helpers.py]: Why do we store the .session and .url if we have the resource?

**Comment by jjdelc** [comment]
Can you add some tests for this please? Maybe create a new ReadOnly subclass on the test file for this and poke at that, or use any of what we have just to make sure this is test covered.

Also, I made a note about a regression/test you removed from the CategoryList.

File: scrunch/datasets.py (hunk 2 above)
Comment (skriems) [inline scrunch/datasets.py]: the reason for that is related to the different approach of accessing the attributes with the new class. Formerly you were using the `fill` method which was directly setting the `Category` attributes (which should be forbidden by now). Changing their attributes with the `edit` method had no effect on the instances in the `CategoryList` so basically there was a conflict between having up to date `Category` instances in the `CategoryList` (`test_edit_category`) with a single intantiation of it (`test_instance_is_reused`).

File: scrunch/helpers.py (hunk 3 above)
Comment (skriems) [inline scrunch/helpers.py]: you're right. Will update and change the code where the session was accessed directly.

**Comment by skriems** [comment]
Yes I was thinking this needs tests as well. Will update tomorrow.

**Pull request merged**